core/domain/stats_jobs_one_off.py

```diff
@@ -298,6 +413,252 @@ def _apply_state_name_changes(cls, prev_stats_dict, change_list_dict):
+                exp_stats_dict = copy.deepcopy(prev_stats_dict)
+
+            # Compute the statistics for events corresponding to this version.
+            state_hit_session_ids, solution_hit_session_ids = set(), set()
+            while event_dict['version'] == version:
+                state_stats = (exp_stats_dict['state_stats_mapping'][
+                    event_dict['state_name']])
+                if event_dict['event_type'] == (
+                        feconf.EVENT_TYPE_START_EXPLORATION):
+                    exp_stats_dict['num_starts_v2'] += 1
+                elif event_dict['event_type'] == (
+                        feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION):
+                    exp_stats_dict['num_actual_starts_v2'] += 1
+                elif event_dict['event_type'] == (
+                        feconf.EVENT_TYPE_COMPLETE_EXPLORATION):
+                    exp_stats_dict['num_completions_v2'] += 1
+                elif event_dict['event_type'] == (
+                        feconf.EVENT_TYPE_ANSWER_SUBMITTED):
+                    state_stats['total_answers_count_v2'] += 1
+                    if event_dict['is_feedback_useful']:
+                        state_stats['useful_feedback_count_v2'] += 1
+                elif event_dict['event_type'] == feconf.EVENT_TYPE_STATE_HIT:
+                    state_stats['total_hit_count_v2'] += 1
+                    state_hit_key = (
+                        event_dict['session_id'] + event_dict['state_name'])
+                    if state_hit_key not in state_hit_session_ids:
+                        state_stats['first_hit_count_v2'] += 1
+                        state_hit_session_ids.add(state_hit_key)
+                elif event_dict['event_type'] == feconf.EVENT_TYPE_SOLUTION_HIT:
+                    solution_hit_key = (
+                        event_dict['session_id'] + event_dict['state_name'])
+                    if solution_hit_key not in solution_hit_session_ids:
+                        state_stats['num_times_solution_viewed_v2'] += 1
+                        solution_hit_session_ids.add(solution_hit_key)
+                elif event_dict['event_type'] == (
+                        feconf.EVENT_TYPE_STATE_COMPLETED):
+                    state_stats['num_completions_v2'] += 1
+                event_dict_idx += 1
+                if event_dict_idx < len(sorted_events_dicts):
+                    event_dict = sorted_events_dicts[event_dict_idx]
+                else:
+                    break
+
+            exp_stats_dicts.append(copy.deepcopy(exp_stats_dict))
+
+        for idx, stats_dict in enumerate(exp_stats_dicts):
+            if stats_dict != old_stats_dicts[idx]:
+                yield (
+                    'ERROR in recomputation. exp_id: %s, exp_version: %s, '
+                    'new_stats_dict: %s' % (exp_id, idx + 1, stats_dict))
+
+    @classmethod
+    def apply_state_name_changes(cls, prev_stats_dict, change_list_dict):
```

Comment:
The purpose of this is broader: upgrade_state_stats_dict perhaps.

Also make it private: prefix with underscore.